src/components/ColorPicker/components/TextPicker/TextPicker.tsx

```diff
@@ -0,0 +1,116 @@
+import React, {useCallback, useEffect, useState} from 'react';
+
+import {classNames} from '../../../../utilities/css';
+import {HSBColor} from '../../../../utilities/color-types';
+import {
+  normalizeColorString,
+  expandHex,
+  hsbToHex,
+  hsbToString,
+  rgbStringToHex,
+  nameToHex,
+  hexToHsb,
+} from '../../../../utilities/color-transformers';
+import {
+  isColorName,
+  isHexString,
+  isHashlessHex,
+  isRgbString,
+} from '../../../../utilities/color-validation';
+import {useI18n} from '../../../../utilities/i18n';
+import {TextField} from '../../../TextField';
+import styles from '../../ColorPicker.scss';
+
+export interface TextPickerProps {
+  color: HSBColor;
+  allowAlpha?: boolean;
+  onChange(hex: string): void;
+}
+
+export function TextPicker({color, allowAlpha, onChange}: TextPickerProps) {
+  const i18n = useI18n();
+
+  const [text, setText] = useState('');
+  const [lastValidValue, setLastValidValue] = useState('');
+
+  const handleTextChange = useCallback((value) => {
+    setText(value);
+  }, []);
+
+  const handleBlur = useCallback(() => {
+    const validUserInput = coerceToValidUserInput(text);
+    if (validUserInput) {
+      setText(validUserInput);
+      setLastValidValue(validUserInput);
+
+      const colorHasChanged = validUserInput !== hsbToHex(color);
+
+      if (colorHasChanged) {
+        onChange(validUserInput);
+      }
+
+      return;
+    }
+
+    setText(lastValidValue);
+  }, [color, lastValidValue, onChange, text]);
+
+  useEffect(() => {
+    const newValue = hsbToHex(color);
+    if (newValue !== hsbToHex(hexToHsb(lastValidValue))) setText(newValue);
+    if (lastValidValue === '') setLastValidValue(newValue);
+  }, [color, lastValidValue]);
+
+  const className = classNames(
+    styles.TextPicker,
+    allowAlpha && styles.AlphaAllowed,
+  );
+  const label = i18n.translate(
+    'Polaris.ColorPicker.textPickerAccessibilityLabel',
+  );
+  const valueForDisplay = isHexString(text) ? text.toUpperCase() : text;
+
+  const renderSelectedColorSwatch = () => {
```

Comment:
Rather than creating a function we usually create a variable and suffix it with markup, i.e

```tsx
const preficClassNames = classNames(
  styles.TextFieldSwatch,
  allowAlpha && styles.AlphaAllowed,
);
const prefixStyle = {backgroundColor: hsbToString(color)};
const prefixMarkup = (
  <div className={prefixClassNames}>
     <div style={prefixStyle} className={styles.SwatchBackground} />
   </div>
)
```